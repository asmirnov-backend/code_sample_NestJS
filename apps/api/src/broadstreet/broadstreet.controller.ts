import {
  Controller,
  Get,
  Query,
  UseInterceptors,
  UsePipes,
  ValidationPipe,
} from '@nestjs/common';

import { LoggingRequestInterceptor } from '@api/logging-request.interceptor';
import { MetadataService } from '@api/metadata/metadata.service';

import { BroadstreetService } from './broadstreet.service';
import { BroadstreetQueryParamsDto } from './dto/broadstreet.params.dto';
import { Broadstreet } from './interfaces/broadstreet.output.interface';

@Controller()
@UseInterceptors(LoggingRequestInterceptor)
export class BroadstreetController {
  constructor(
    private readonly broadstreet: BroadstreetService,
    private readonly metadata: MetadataService,
  ) {}

  @Get('/broadstreet')
  @UsePipes(new ValidationPipe({ transform: true }))
  async getBroadstreet(
    @Query() query: BroadstreetQueryParamsDto,
  ): Promise<Broadstreet> {
    const advertiser = await this.metadata.getAdvertiser(query.advertiserId);

    const campaigns = await this.metadata.getCampaigns({
      advertiserId: advertiser.id,
      campaignOriginIds: query.campaignIds,
    });

    return this.broadstreet.getBroadstreet({
      campaigns: campaigns,
    });
  }
}
