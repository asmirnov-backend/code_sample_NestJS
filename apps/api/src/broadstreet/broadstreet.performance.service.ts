import { Injectable } from '@nestjs/common';
import { DateTime } from 'luxon';

import { PrismaService } from '@api/prisma/prisma.service';
import { PerformanceParams } from '@common/interfaces/performance.params.interface';
import { calculateDatesForPerformanceByDay } from '@common/utils/calculate-dates';

import {
  BroadstreetPerformance,
  BroadstreetPerformanceByCreatives,
} from './interfaces/broadstreet.performance.interface';

@Injectable()
export class BroadstreetPerformanceService {
  constructor(private readonly prisma: PrismaService) {}

  public async getBroadstreetPerformanceCommon(
    params: PerformanceParams,
  ): Promise<BroadstreetPerformance[]> {
    const performance =
      await this.prisma.broadstreet_performance_campaign.groupBy({
        by: ['date'],
        where: {
          campaign_id: {
            in: params.campaignIds,
          },
          date: calculateDatesForPerformanceByDay(params),
        },
        _sum: {
          views: true,
          hovers: true,
          clicks: true,
        },
      });

    return performance.map((e) => ({
      date: DateTime.fromJSDate(e.date),
      impressions: e._sum.views ?? 0,
      clicks: e._sum.clicks ?? 0,
      hovers: e._sum.hovers ?? 0,
    }));
  }

  public async getBroadstreetPerformanceByCreatives(
    params: PerformanceParams,
  ): Promise<BroadstreetPerformanceByCreatives[]> {
    const performance =
      await this.prisma.broadstreet_performance_creatives.groupBy({
        by: ['creative_id', 'creative_name'],
        where: {
          campaign_id: {
            in: params.campaignIds,
          },
          date: calculateDatesForPerformanceByDay(params),
        },
        _sum: {
          views: true,
          hovers: true,
          clicks: true,
        },
      });

    return performance.map((e) => ({
      creativeId: e.creative_id,
      creativeName: e.creative_name,
      impressions: e._sum.views ?? 0,
      clicks: e._sum.clicks ?? 0,
      hovers: e._sum.hovers ?? 0,
    }));
  }
}
