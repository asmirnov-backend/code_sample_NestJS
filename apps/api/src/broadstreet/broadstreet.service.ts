import { BroadstreetPerformanceService } from './broadstreet.performance.service';
import {
  Broadstreet,
  BroadstreetTotal,
  BroadstreetByDayOfWeek,
  BroadstreetByDate,
  BroadstreetByCreative,
} from './interfaces/broadstreet.output.interface';

import { DATE_FORMAT } from '@common/consts/date-format.const';
import { CampaignsWithInterval } from '@common/interfaces/campaign-with-interval.interface';
import { calcPercentage } from '@common/utils/calc-percentage';
import { getSpendData } from '@common/utils/get-spend-data';

import { Injectable } from '@nestjs/common';
import { AnalyticsType } from '@prisma/client';
import { groupBy, map, max, min, sumBy } from 'lodash';

@Injectable()
export class BroadstreetService {
  constructor(private readonly performance: BroadstreetPerformanceService) {}

  public async getBroadstreet(
    params: CampaignsWithInterval,
  ): Promise<Broadstreet> {
    return {
      TabName: AnalyticsType.BROADSTREET,
      DayOfWeekImpression: await this.getBroadstreetByDayOfWeek(params),
      ByDateImpression: await this.getBroadstreetByDate(params),
      CreativeList: await this.getBroadstreetByCreative(params),
    };
  }

  public async getBroadstreetTotal(
    params: CampaignsWithInterval,
  ): Promise<BroadstreetTotal> {
    const performance = await this.performance.getBroadstreetPerformanceCommon({
      campaignIds: params.campaigns.map((e) => e.originId),
      interval: params.interval,
    });

    const impressions = sumBy(performance, 'impressions');
    const hovers = sumBy(performance, 'hovers');
    const clicks = sumBy(performance, 'clicks');
    const ctr = calcPercentage(clicks, impressions);

    const spendData = getSpendData(
      impressions,
      params.campaigns.map((e) => ({
        spendRate: Number(e.spendRate),
        spendType: e.spendType,
      })),
    );

    const dates = map(performance, 'date');
    const start = min(dates)?.toFormat(DATE_FORMAT);
    const end = max(dates)?.toFormat(DATE_FORMAT);

    return {
      AnalyticsType: AnalyticsType.BROADSTREET,
      Impressions: impressions,
      StartDate: start,
      EndDate: end,
      Hovers: hovers,
      Clicks: clicks,
      CTR: ctr,
      Spend: spendData.spend,
      SpendVal: spendData.spendVal,
    };
  }

  public async getBroadstreetByDayOfWeek(
    params: CampaignsWithInterval,
  ): Promise<BroadstreetByDayOfWeek[]> {
    const performance = await this.performance.getBroadstreetPerformanceCommon({
      campaignIds: params.campaigns.map((e) => e.originId),
      interval: params.interval,
    });

    const allImpressions = sumBy(performance, 'impressions');
    const groupByWeekday = groupBy(
      performance,
      (row) => row.date.setLocale('en').weekdayLong,
    );

    return map(groupByWeekday, (data, weekday) => {
      const impressions = sumBy(data, 'impressions');
      const clicks = sumBy(data, 'clicks');

      return {
        AnalyticsType: AnalyticsType.BROADSTREET,
        Day: weekday,
        Impressions: impressions,
        Clicks: clicks,
        Hovers: sumBy(data, 'hovers'),
        ImpressionPercent: calcPercentage(impressions, allImpressions),
        CTR: calcPercentage(clicks, impressions),
      };
    });
  }

  public async getBroadstreetByDate(
    params: CampaignsWithInterval,
  ): Promise<BroadstreetByDate[]> {
    const performance = await this.performance.getBroadstreetPerformanceCommon({
      campaignIds: params.campaigns.map((e) => e.originId),
      interval: params.interval,
    });

    return performance.map((row) => ({
      AnalyticsType: AnalyticsType.BROADSTREET,
      Date: row.date.toFormat(DATE_FORMAT),
      Impressions: row.impressions,
      Clicks: row.clicks,
      Hovers: row.hovers,
      CTR: calcPercentage(row.clicks, row.impressions),
    }));
  }

  public async getBroadstreetByCreative(
    params: CampaignsWithInterval,
  ): Promise<BroadstreetByCreative[]> {
    const performance =
      await this.performance.getBroadstreetPerformanceByCreatives({
        campaignIds: params.campaigns.map((e) => e.originId),
        interval: params.interval,
      });

    return performance.map((row) => ({
      CreativeId: row.creativeId,
      CreativeName: row.creativeName,
      Impressions: row.impressions,
      Hovers: row.hovers,
      Clicks: row.clicks,
      CTR: calcPercentage(row.clicks, row.impressions),
    }));
  }
}
