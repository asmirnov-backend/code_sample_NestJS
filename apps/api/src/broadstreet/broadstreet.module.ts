import { Module } from '@nestjs/common';

import { BroadstreetController } from './broadstreet.controller';
import { BroadstreetPerformanceService } from './broadstreet.performance.service';
import { BroadstreetService } from './broadstreet.service';

import { PrismaModule } from '../prisma/prisma.module';

@Module({
  imports: [PrismaModule],
  providers: [
    BroadstreetService,
    BroadstreetPerformanceService,
    BroadstreetController,
  ],
})
export class BroadstreetModule {}
