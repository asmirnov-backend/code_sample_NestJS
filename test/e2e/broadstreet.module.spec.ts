import { INestApplication } from '@nestjs/common';
import { ConfigModule } from '@nestjs/config';
import { TestingModule, Test } from '@nestjs/testing';
import * as request from 'supertest';

import { BroadstreetModule } from '@api/broadstreet/broadstreet.module';
import { Broadstreet } from '@api/broadstreet/interfaces/broadstreet.output.interface';

jest.setTimeout(10_000);

describe('BroadstreetModule', () => {
  let app: TestingModule;
  let nestApp: INestApplication;
  let server;

  beforeAll(async () => {
    app = await Test.createTestingModule({
      imports: [ConfigModule.forRoot({ isGlobal: true }), BroadstreetModule],
    }).compile();

    nestApp = app.createNestApplication();
    await nestApp.init();
    server = nestApp.getHttpServer();
  });

  afterAll(async () => await nestApp.close());

  it('/broadstreet', async () => {
    const result = await request(server).get(
      '/broadstreet?advertiserId=1&campaignIds=1',
    );

    const data: Broadstreet = result.body;
    expect(data).toMatchSnapshot();
  });
});
