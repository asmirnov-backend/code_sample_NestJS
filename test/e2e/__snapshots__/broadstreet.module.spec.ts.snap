// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BroadstreetModule /broadstreet 1`] = `
Object {
  "ByDateImpression": Array [
    Object {
      "AnalyticsType": "BROADSTREET",
      "CTR": "10.00%",
      "Clicks": 1,
      "Date": "2022-01-01",
      "Hovers": 1,
      "Impressions": 10,
    },
    Object {
      "AnalyticsType": "BROADSTREET",
      "CTR": "10.00%",
      "Clicks": 1,
      "Date": "2022-01-02",
      "Hovers": 1,
      "Impressions": 10,
    },
  ],
  "CreativeList": Array [
    Object {
      "CTR": "10.00%",
      "Clicks": 1,
      "CreativeId": 1,
      "CreativeName": "creative_name 1",
      "Hovers": 1,
      "Impressions": 10,
    },
    Object {
      "CTR": "10.00%",
      "Clicks": 2,
      "CreativeId": 2,
      "CreativeName": "creative_name 2",
      "Hovers": 2,
      "Impressions": 20,
    },
  ],
  "DayOfWeekImpression": Array [
    Object {
      "AnalyticsType": "BROADSTREET",
      "CTR": "10.00%",
      "Clicks": 1,
      "Day": "Saturday",
      "Hovers": 1,
      "ImpressionPercent": "50.00%",
      "Impressions": 10,
    },
    Object {
      "AnalyticsType": "BROADSTREET",
      "CTR": "10.00%",
      "Clicks": 1,
      "Day": "Sunday",
      "Hovers": 1,
      "ImpressionPercent": "50.00%",
      "Impressions": 10,
    },
  ],
  "TabName": "BROADSTREET",
}
`;
